### PR TITLE
[Diagnostics] Port explicit closure result contextual mismatch

### PR DESCRIPTION
**Pull request by xedin**
Detect and diagnose a problem when explicitly specified closure
result type doesn't match what is expected by the context:

Example:

```swift
func foo(_: () -> Int) {}

foo { () -> String in "" } // `Int` vs. `String`
```

### CODE REVIEW

**Comment by xedin** [comment]
This is also related to `diagnoseGeneral*Failure` removal from CSDiag.

**Comment by xedin** [comment]
@swift-ci please smoke test

**Review by hborla — APPROVED** [review]

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

**Comment by xedin** [comment]
@swift-ci please smoke test macOS platform

File: test/Sema/immutability.swift [hunk 1]
```diff
@@ -656,8 +656,8 @@ func sr4214() {
   let closure = { val in val.x = 7 } as (inout MutableSubscripts) -> () // Ok
   var v = MutableSubscripts()
   closure(&v)
-  // FIXME: This diagnostic isn't really all that much better
-  // expected-error@+1 {{cannot convert value of type '(inout MutableSubscripts) -> ()' to expected argument type '(_) -> _'}}
```
Comment (slavapestov) [inline test/Sema/immutability.swift]: Nice!

**Pull request merged**